java-checks/src/main/java/org/sonar/java/checks/helpers/SupersetAutomataChecker.java

```diff
@@ -31,13 +35,20 @@ protected boolean neutralAnswer() {
 
   @Override
   protected boolean checkAuto1AndAuto2Successors(SubAutomaton auto1, SubAutomaton auto2, boolean defaultAnswer, boolean hasConsumedInput) {
-    SimplifiedRegexCharacterClass characterClass1 = SimplifiedRegexCharacterClass.of(auto1.start);
-    SimplifiedRegexCharacterClass characterClass2 = SimplifiedRegexCharacterClass.of(auto2.start);
-    return ((characterClass1 != null) && (characterClass2 != null)) ?
-      (characterClass1.supersetOf(characterClass2, defaultAnswer) &&
-        auto2.allSuccessorMatch(successor2 -> auto1.anySuccessorMatch(successor1 ->
-          check(successor1, successor2, true)))) :
-      defaultAnswer;
+    boolean start1SupersetOfStart2;
+    if (auto1.incomingTransitionType() == BOUNDARY && auto2.incomingTransitionType() == BOUNDARY) {
+      start1SupersetOfStart2 = ((BoundaryTree) auto1.start).type() == ((BoundaryTree) auto2.start).type();
```

Comment:
Here we should use `defaultAnswer` instead of `false` when the types are not equal. Alternatively, we can implement a partial hierarchy for boundaries like this:

UNICODE_EXTENDED_GRAPHEME_CLUSTER ⊇WORD ⊇ LINE_START ⊇ INPUT_START
UNICODE_EXTENDED_GRAPHEME_CLUSTER ⊇WORD ⊇ LINE_END ⊇ INPUT_END_FINAL_TERMINATOR ⊇INPUT_END
NON_WORD isn't a superset or subset of any other boundary.